pkg/autodiscovery/listeners/kubelet_containermeta.go

```diff
@@ -0,0 +1,311 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+// +build kubelet
+
+// TODO(juliogreff): remove build tags after we remove kubelet listener and
+// move constants here
+
+package listeners
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/common/utils"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/containermeta"
+	"github.com/DataDog/datadog-agent/pkg/status/health"
+	"github.com/DataDog/datadog-agent/pkg/util/containers"
+	"github.com/DataDog/datadog-agent/pkg/util/kubernetes/kubelet"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+func init() {
+	Register("kubelet-containermeta", NewKubeletContainerMetaListener)
+}
+
+// KubeletContainerMetaListener listens to pod creation through a subscription
+// to the containermeta store.
+type KubeletContainerMetaListener struct {
+	store *containermeta.Store
+	stop  chan struct{}
+
+	mu       sync.RWMutex
+	filters  *containerFilters
+	services map[string]Service
+
+	newService chan<- Service
+	delService chan<- Service
+}
+
+// NewKubeletContainerMetaListener returns a new KubeletContainerMetaListener.
+func NewKubeletContainerMetaListener() (ServiceListener, error) {
+	filters, err := newContainerFilters()
+	if err != nil {
+		return nil, err
+	}
+
+	return &KubeletContainerMetaListener{
+		store:    containermeta.GlobalStore,
+		filters:  filters,
+		services: make(map[string]Service),
+		stop:     make(chan struct{}),
+	}, nil
+}
+
+// Listen starts listening to events from the containermeta store.
+func (l *KubeletContainerMetaListener) Listen(newSvc chan<- Service, delSvc chan<- Service) {
+	l.newService = newSvc
+	l.delService = delSvc
+
+	const name = "ad-containermeta-kubeletlistener"
+
+	ch := l.store.Subscribe(name, &containermeta.Filter{
+		Sources: []string{"kubelet"},
```

Comment:
It might not be necessary, but I'm not 100% sure yet. We'll see when we add the kubemetadata collector (so pods come from several sources), and if it makes a difference in AD if a container comes from kubelet or docker collectors. If it ends up being all the same, I'll be happy to rip this out.